l2-contracts/contracts/ConsensusRegistry.sol

```diff
@@ -0,0 +1,278 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.20;
+
+import {Ownable2Step} from "@openzeppelin/contracts/access/Ownable2Step.sol";
+import {EfficientCall} from "@matterlabs/zksync-contracts/l2/system-contracts/libraries/EfficientCall.sol";
+
+/// @author Matter Labs
+/// @custom:security-contact security@matterlabs.dev
+/// @title ConsensusRegistry
+/// @dev Manages consensus nodes and committees for the L2 consensus protocol,
+/// owned by a governance protocol. Nodes act as both validators and attesters,
+/// each playing a distinct role in the consensus process. This contract facilitates
+/// the rotation of validator and attester committees, which represent a subset of nodes
+/// expected to actively participate in the consensus process during a specific time window.
+contract ConsensusRegistry is Ownable2Step {
+    // An array to keep track of node owners.
+    address[] public nodeOwners;
+    // A map of node owners => nodes.
+    mapping(address => Node) public nodes;
+    // The current validator committee list.
+    CommitteeValidator[] public validatorCommittee;
+    // The current attester committee list.
+    CommitteeAttester[] public attesterCommittee;
+
+    /// @dev Represents a consensus node.
+    struct Node {
```

Comment:
The node data can be easily packed by changing types and order.

validatorPoP - split into three values bytes32, bytes32, bytes32
attesterPubKey - bytes32,
validatorPubKey - split into two values bytes32 and bytes16
bool active - bool,
attesterWeight and validatorWeight can be smaller, maybe uint32? Isn't this enough?


This way the storage consumption will drop from 12 storage slots to 6 (because dynamic data types stores 1 word for size + optimal packing saves other 2-3 words)

Reply:
The reason I didn’t use fixed sizes was to avoid committing to a specific scheme. This can indeed be optimized.

* Validator public key is a `bls12_381` p2: 96 bytes
* Validator proof of possession (signature) is a `bls12_381` p1: 48 bytes
* Attester public key is a secp256k1 point: 33 bytes

Additionally, I wasn’t sure if we would end up using the user-friendly representation, which declares the type (for potential validation). For example:
```
attester:public:secp256k1:0320d5f3d93310c290ef40c51e892e0386aa9feefc91e3196af56df51b15378dc5
```

As for the weights, I assume uint32 is indeed sufficient.

@brunoffranca @pompon0 any thoughts?

Reply:
We can use fixed sizes. The schemes are not expected to change that often. Also no point in using the qualifiers, it's clear from context what each key is. We only use qualifiers for human-readable text.

Reply:
Applied here: https://github.com/matter-labs/era-contracts/pull/555/commits/a5b796caa52ef24cf3e385f4199e17bf506ac5bb (test update to follow)